crates/client/state-sync/src/ethereum.rs

```diff
@@ -0,0 +1,668 @@
+            .ok_or(Error::Other("query transaction by hash get none".to_string()))?;
+
+        let abi = BaseContract::from(
+            parse_abi(&["function registerContinuousMemoryPage(uint256 startAddr,uint256[] values,uint256 z,uint256 \
+                         alpha,uint256 prime)"])
+            .unwrap(),
+        );
+
+        let (_, data, _, _, _): (U256, Vec<U256>, U256, U256, U256) =
+            abi.decode("registerContinuousMemoryPage", tx.input.as_ref()).unwrap();
+        Ok(data)
+    }
+
+    /// Decodes the StarkNet state diff from raw data.
+    ///
+    /// # Arguments
+    /// * `l1_block_number` - The L1 block number.
+    /// * `starknet_block_number` - The StarkNet block number.
+    /// * `data` - The raw data containing the state diff information.
+    /// * `client` - The StarkNet runtime client.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the decoded `StateDiff` or an `Error` if decoding fails.
+    pub fn decode_state_diff<B, C>(
+        &self,
+        starknet_block_number: u64,
+        data: Vec<U256>,
+        client: Arc<C>,
+    ) -> Result<StateDiff, Error>
+    where
+        B: BlockT,
+        C: ProvideRuntimeApi<B> + HeaderBackend<B>,
+        C::Api: StarknetRuntimeApi<B>,
+    {
+        if starknet_block_number < self.v011_diff_format_height {
+            let with_constructor_args = starknet_block_number < self.constructor_args_diff_height;
+            parser::decode_pre_011_diff(&data, with_constructor_args)
+        } else {
+            let block_hash = client
+                .block_hash_from_id(&BlockId::Number((starknet_block_number as u32).saturating_sub(1).into()))
+                .map_err(|_| Error::UnknownBlock)?
+                .unwrap_or_default();
+            parser::decode_011_diff(&data, block_hash, client)
+        }
+    }
+
+    /// Queries Ethereum state differences for a specific StarkNet state update.
+    ///
+    /// # Arguments
+    ///
+    /// * `state_update` - The state update containing Ethereum origin and log state update data.
+    /// * `client` - The StarkNet runtime client.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the fetched `FetchState` or an `Error` if the query fails.
+    pub async fn query_state_diff<B, C>(
+        &mut self,
+        state_update: &StateUpdate,
+        client: Arc<C>,
+    ) -> Result<FetchState, Error>
+    where
+        B: BlockT,
+        C: ProvideRuntimeApi<B> + HeaderBackend<B>,
+        C::Api: StarknetRuntimeApi<B>,
+    {
+        debug!(target: LOG_TARGET,"~~ query state diff for starknet block {:#?}", state_update.update.block_number);
+
+        let fact = self
+            .query_state_transition_fact(
+                state_update.eth_origin.block_number,
+                state_update.eth_origin.transaction_index,
+            )
+            .await?;
+
+        let pages_hashes = self.query_memory_pages_hashes(state_update.eth_origin.block_number, fact).await?;
```

Comment:
There is no reason those two could not be run in parallel.

Reply:
They should not run in parallel; the first one is searching for a fact, and the second one depends on this fact.

Reply:
Oh sorry I missed that